beacon-chain/light-client-network.md

```diff
@@ -0,0 +1,200 @@
+
+#### Validation
+Validating `LightClientFinalityUpdate` and `LightClientOptimisticUpdate` follows the gossip domain(gossipsub) [consensus specs](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/p2p-interface.md#the-gossip-domain-gossipsub).
+
+### Data
+
+#### Types
+
+* LightClientBootstrap
+* LightClientUpdate
+* LightClientFinalityUpdate
+* LightClientOptimisticUpdate
+
+All data types are specified in light client [sync protocol](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md#containers).
+
+#### Retrieval
+
+* Requests the `LightClientBootstrap` structure corresponding to a given post-Altair beacon block root.
+* `LightClientUpdatesByRange` - requests the `LightClientUpdate` instances in the sync committee period range [start_period, start_period + count), leading up to the current head sync committee period as selected by fork choice.
+* The latest `LightClientFinalityUpdate` known by a peer.
+* The latest `LightClientOptimisticUpdate` known by a peer.
+
+## Specification
+
+### Distance
+
+Nodes in the light client network are represented by their [EIP-778 Ethereum Node Record (ENR)](https://eips.ethereum.org/EIPS/eip-778) from the Discovery v5 network.
+A node's `node-id` is derived according to the node's identity scheme, which is specified in the node's ENR. A node's `node-id` represents its address in the DHT.
+
+The `node-id` is a 32-byte identifier. We define the `distance` function that maps a pair of `node-id` values to a 256-bit unsigned integer identically to the Discovery v5 network.
+
+```
+distance(n1, n2) = n1 XOR n2
+```
+
+Similarly, we define a `logdistance` function identically to the Discovery v5 network.
+
+```
+logdistance(n1, n2) = log2(distance(n1, n2))
+```
+
+### Content: Keys and Values
+
+The beacon chain light client DHT stores the following data items:
+
+* LightClientBootstrap
+* LightClientUpdate
+* LightClientFinalityUpdate
+* LightClientOptimisticUpdate
+
+Each of these data items are represented as a key-value pair.
+
+- The "key" for each data item is defined as `content_key`.
+- The "value" for each data item is defined as `content`.
+
+See each of the individual data item definitions for their individual `content` and `content_key` definitions.
+
+#### Constants
+
+```py
+# Maximum number of `LightClientUpdate` instances in a single request
+MAX_REQUEST_LIGHT_CLIENT_UPDATES = 2**7  # = 128
+```
+
+#### LightClientBootstrap
+
+```
+light_client_bootstrap_key = Container(block_hash: Bytes32)
+selector                   = 0x00
+
+content                    = SSZ.serialize(LightlientBootstrap)
+content_key                = selector + SSZ.serialize(light_client_bootstrap_key)
+```
+
+#### LightClientUpdatesByRange
+
+```
+light_client_update_keys   = Container(start_period: uint64, count: uint64)
```

Comment:
I think these need to be fixed ranges in order to allow for unique content ids that than match radii of nodes?

Or, nodes could store all data, see bigger comment.

Reply:
@kdeme @ogenev what your opinion on having this as:
`light_client_update_key   = Container(period: uint64)`

So each update will have each own content key and content id ?

I think one of the problem with each peer storing everything is the fact that there maybe cases when you will get stuck because closest peer would have less that desirable number of updates. As example:
1. lets say peer closest to local node will have 2 updates
2. now when issuing `findContent` for all updates we will always receive this 2 updates even though we may need more. It will happen unless this closes peer updates its view (receives more updates) or we get rid of him from our routing table.

Having each update under different key does not have this problem. I think main problem with this approach is that node needs to issue lookup for each key which have bigger latency but:
- its not big a deal as each period lasts 27h so it is pretty long, so even with higher latency we won't hit the case that we can't catch up with the network. 
- we can always do few concurrent lookups i.e light client network can have interface `getLightClientUpdatesByRange(start_period: uint64, count: uint64)` which underneath will translate this query to several concurrent queries and aggregates results. (or stream results as they come)

Reply:
> I think one of the problem with each peer storing everything is the fact that there maybe cases when you will get stuck because closest peer would have less that desirable number of updates. As example:

Lookups will/must request to more than just 1 neighboring peer. Else, same issue can occur for just other type of data too. 
That being said, in the scenario that every peer stores everything, one could just randomly ask peers, similar as we used to do for the "latest" master accumulator (which has been removed now) in the history network.



> * its not big a deal as each period lasts 27h so it is pretty long, so even with higher latency we won't hit the case that we can't catch up with the network.

It is not just about just catching up. I think in Portal clients it will be important to know the head of the chain as fast as possible. As other networks / functionality will rely on this. So for a good UX, I think it will be important that this happens fast (what is fast enough? Lets find out..). I'm assuming that in the current libp2p version the max count of 128 is used as much as possible? Could do the test with lower count values.

Reply:
One could argue that most important thing to quickly sync up to the tip is to choose bootstrap which is as close to the tip as possible, then whole dance around light client updates is not really that important. (not sure how often those are made)
 
> I'm assuming that in the current libp2p version the max count of 128 is used as much as possible? 

yup that is the case. There is check comparing what is current local sync comitee period and what is latest period, and requesting `min(128, latest - current_local)`

> As other networks / functionality will rely on this. So for a good UX, I think it will be important that this happens fast (what is fast enough? Lets find out..).

I kinda agree on UX point, but at this point it is just guessing game. I also think that distributing 20mb of data across DHT is kinda weird. But on the other hand I find model `1 content item <-> 1 content key` much more naturally fitting DHT, and a bit more flexible (nothing prevents node to asks for multiple updates concurrently), where having this `1 contentKey <-> all items` feels a bit clunky where one need to resolve what does it really mean all items.

Ultimately, one could have both:
- contentKey to ask for all updates
- contentKey to ask for specific update per period.
Then peer first ask for all updates to some random peers, and after that, if necessary ask for specific updates which weren't provided by those peers.
This probably complicates implementation a bit, but we have best of two worlds.

Reply:
Another slight problem with having only option of asking for all updates is that  in case where some peer starts from recent bootstrap, and it request all updates starting from this bootstrap to most recent period, then it may receive a lot of all light client updates which are of no use and which cannot be validated in any way.

As example: 
- I am starting from period 510 (thats where I go my trusted bootstrap), the most recent one is 528
- I am asking for all updates
- I may receive all light client updates from 400 to 528 ( as this is max 128 updates that should be stored)
- So now I presented with a problem: should I discard almost 110 of updates (as I do not have a way of validating them) or should I try to get boostrap for period 400. If I discard them it is a lot of wasted bandwidth. I if decide to get older bootstrap, how do I make sure I trust it. (i.e user have chosen different trusted bootstrap, should our software make decision for user to trust some other one ? probably not)

I imagine there is way around it, by changing our arbitrary data in ping/pong to cary info about oldest stored period  by some peer (as we assume every peer stores everything, up to weak subjectivty boundary, in this network we probably do not need concept of radius) and chose to peer which the best fit our bill i.e his starting period is closest to our starting bootstrap.

Reply:
> * contentKey to ask for all updates

When I brought up the possible idea of storing all updates (with a purging after updates are older than the required time to store), I didn't mean to also send them all over a request with 1 `contentKey`.

It was as a solution to be able to do the range requests. So a the `contentKey` would be related to the range requested.

But it is not the only solution. We could also just hard set certain ranges that can be requested, e.g fixed start period + fixed range (128?).

If the ranges are required in the first place of course (for good UX). I realize that this depends on how old the bootstrap is.
The question here is how this would work exactly in a real scenario such as a wallet using a Portal client as back-end.

Reply:
> When I brought up the possible idea of storing all updates (with a purging after updates are older than the required time to store), I didn't mean to also send them all over a request with 1 contentKey.

yea, I think initially I have misunderstood this point 😅  

But then having content key as `Container(start_period: uint64, count: uint64))` and every node storing everything, we most probably need some custom lookup procedure as content id from such content key does not have any meaning really. 

Thats why I proposed in my last comment that maybe it would be valuable, that each node insead of radius (as if we store everything it does not have any value) would broadcast oldest sync comitee period for which it has data available, so that we can lookup for peers and determine which has data from period intersting to us.

> The question here is how this would work exactly in a real scenario such as a wallet using a Portal client as back-end.

yup, my current metal model is that user would only need to specify trusted bootstrap and sync form boostrap only if he was offline longer than 5 months, other than that progress would be saved and lc would be restared from most recent sync comittee period.